File: src/graphql/mutations/billing/payments.ts
import { graphql } from "@/lib/graphql/client";

// ============================================================================
// PAYMENT MUTATIONS
// ============================================================================

/**
 * Creates a new billing payment.
 * @param payload The input data for creating the payment.
 */
export const createPayment = graphql(`
  mutation CreatePayment($payload: CreatePaymentInput!) {
    billing {
      createPayment(payload: $payload) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the invoice ID of a billing payment.
 * @param id The UUID of the payment to update.
 * @param invoiceId The new invoice ID.
 */
export const updatePaymentInvoiceId = graphql(`
  mutation UpdatePaymentInvoiceId($id: UUID!, $invoiceId: UUID!) {
    billing {
      updatePaymentInvoiceId(id: $id, invoiceId: $invoiceId) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the amount of a billing payment.
 * @param id The UUID of the payment to update.
 * @param amount The new amount.
 */
export const updatePaymentAmount = graphql(`
  mutation UpdatePaymentAmount($id: UUID!, $amount: Float!) {
    billing {
      updatePaymentAmount(id: $id, amount: $amount) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the payment method of a billing payment.
 * @param id The UUID of the payment to update.
 * @param paymentMethod The new payment method.
 */
export const updatePaymentPaymentMethod = graphql(`
  mutation UpdatePaymentPaymentMethod($id: UUID!, $paymentMethod: PaymentMethodEnum!) {
    billing {
      updatePaymentPaymentMethod(id: $id, paymentMethod: $paymentMethod) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the transaction ID of a billing payment.
 * @param id The UUID of the payment to update.
 * @param transactionId The new transaction ID.
 */
export const updatePaymentTransactionId = graphql(`
  mutation UpdatePaymentTransactionId($id: UUID!, $transactionId: String) {
    billing {
      updatePaymentTransactionId(id: $id, transactionId: $transactionId) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the gateway reference of a billing payment.
 * @param id The UUID of the payment to update.
 * @param gatewayReference The new gateway reference.
 */
export const updatePaymentGatewayReference = graphql(`
  mutation UpdatePaymentGatewayReference($id: UUID!, $gatewayReference: String) {
    billing {
      updatePaymentGatewayReference(id: $id, gatewayReference: $gatewayReference) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the status of a billing payment.
 * @param id The UUID of the payment to update.
 * @param status The new status.
 */
export const updatePaymentStatus = graphql(`
  mutation UpdatePaymentStatus($id: UUID!, $status: PaymentStatusEnum!) {
    billing {
      updatePaymentStatus(id: $id, status: $status) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the payment date of a billing payment.
 * @param id The UUID of the payment to update.
 * @param paymentDate The new payment date.
 */
export const updatePaymentPaymentDate = graphql(`
  mutation UpdatePaymentPaymentDate($id: UUID!, $paymentDate: DateTime) {
    billing {
      updatePaymentPaymentDate(id: $id, paymentDate: $paymentDate) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the processed at timestamp of a billing payment.
 * @param id The UUID of the payment to update.
 * @param processedAt The new processed at timestamp.
 */
export const updatePaymentProcessedAt = graphql(`
  mutation UpdatePaymentProcessedAt($id: UUID!, $processedAt: DateTime) {
    billing {
      updatePaymentProcessedAt(id: $id, processedAt: $processedAt) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the currency of a billing payment.
 * @param id The UUID of the payment to update.
 * @param currency The new currency.
 */
export const updatePaymentCurrency = graphql(`
  mutation UpdatePaymentCurrency($id: UUID!, $currency: String) {
    billing {
      updatePaymentCurrency(id: $id, currency: $currency) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the exchange rate of a billing payment.
 * @param id The UUID of the payment to update.
 * @param exchangeRate The new exchange rate.
 */
export const updatePaymentExchangeRate = graphql(`
  mutation UpdatePaymentExchangeRate($id: UUID!, $exchangeRate: Float) {
    billing {
      updatePaymentExchangeRate(id: $id, exchangeRate: $exchangeRate) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the fees of a billing payment.
 * @param id The UUID of the payment to update.
 * @param fees The new fees.
 */
export const updatePaymentFees = graphql(`
  mutation UpdatePaymentFees($id: UUID!, $fees: Float) {
    billing {
      updatePaymentFees(id: $id, fees: $fees) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the notes of a billing payment.
 * @param id The UUID of the payment to update.
 * @param notes The new notes.
 */
export const updatePaymentNotes = graphql(`
  mutation UpdatePaymentNotes($id: UUID!, $notes: String) {
    billing {
      updatePaymentNotes(id: $id, notes: $notes) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Updates the processed by user ID of a billing payment.
 * @param id The UUID of the payment to update.
 * @param processedByUserId The new processed by user ID.
 */
export const updatePaymentProcessedByUserId = graphql(`
  mutation UpdatePaymentProcessedByUserId($id: UUID!, $processedByUserId: UUID) {
    billing {
      updatePaymentProcessedByUserId(id: $id, processedByUserId: $processedByUserId) {
        id
        amount
        paymentMethod
        transactionId
        gatewayReference
        status
        paymentDate
        processedAt
        currency
        exchangeRate
        fees
        netAmount
        notes
        createdAt
        updatedAt
        invoice {
          id
          invoiceNumber
        }
        processedByUser {
          id
          name
        }
      }
    }
  }
`);

/**
 * Removes a billing payment by its ID.
 * @param id The UUID of the payment to remove.
 */
export const removePayment = graphql(`
  mutation RemovePayment($id: UUID!) {
    billing {
      removePayment(id: $id)
    }
  }
`);

File: services/graphql-billing/src/mutation/payments.rs
use async_graphql::{Context, InputObject, Object};
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::models::{
    enums::{PaymentMethodEnum, PaymentStatusEnum},
    payments,
};

#[derive(Debug, Clone, InputObject)]
pub struct CreatePaymentInput {
    pub invoice_id: Uuid,
    pub amount: f64,
    pub payment_method: PaymentMethodEnum,
    pub transaction_id: Option<String>,
    pub gateway_reference: Option<String>,
    pub status: Option<PaymentStatusEnum>,
    pub payment_date: Option<DateTime<Utc>>,
    pub processed_at: Option<DateTime<Utc>>,
    pub currency: Option<String>,
    pub exchange_rate: Option<f64>,
    pub fees: Option<f64>,
    pub notes: Option<String>,
    pub processed_by_user_id: Option<Uuid>,
}

#[derive(Debug, Clone, Default)]
pub struct PaymentsMutation;

#[Object(name = "BillingPaymentsMutation")]
impl PaymentsMutation {
    async fn create_payment(
        &self,
        ctx: &Context<'_>,
        payload: CreatePaymentInput,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "insert into billing.payments (invoice_id, amount, payment_method, transaction_id, gateway_reference, status, payment_date, processed_at, currency, exchange_rate, fees, notes, processed_by_user_id) values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) returning *",
        )
        .bind(payload.invoice_id)
        .bind(payload.amount)
        .bind(payload.payment_method)
        .bind(payload.transaction_id)
        .bind(payload.gateway_reference)
        .bind(payload.status)
        .bind(payload.payment_date)
        .bind(payload.processed_at)
        .bind(payload.currency)
        .bind(payload.exchange_rate)
        .bind(payload.fees)
        .bind(payload.notes)
        .bind(payload.processed_by_user_id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_invoice_id(
        &self,
        ctx: &Context<'_>,
        invoice_id: Uuid,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set invoice_id = $1 where id = $2 returning *",
        )
        .bind(invoice_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_amount(
        &self,
        ctx: &Context<'_>,
        amount: f64,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set amount = $1 where id = $2 returning *",
        )
        .bind(amount)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_payment_method(
        &self,
        ctx: &Context<'_>,
        payment_method: PaymentMethodEnum,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set payment_method = $1 where id = $2 returning *",
        )
        .bind(payment_method)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_transaction_id(
        &self,
        ctx: &Context<'_>,
        transaction_id: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set transaction_id = $1 where id = $2 returning *",
        )
        .bind(transaction_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_gateway_reference(
        &self,
        ctx: &Context<'_>,
        gateway_reference: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set gateway_reference = $1 where id = $2 returning *",
        )
        .bind(gateway_reference)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_status(
        &self,
        ctx: &Context<'_>,
        status: PaymentStatusEnum,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set status = $1 where id = $2 returning *",
        )
        .bind(status)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_payment_date(
        &self,
        ctx: &Context<'_>,
        payment_date: Option<DateTime<Utc>>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set payment_date = $1 where id = $2 returning *",
        )
        .bind(payment_date)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_processed_at(
        &self,
        ctx: &Context<'_>,
        processed_at: Option<DateTime<Utc>>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set processed_at = $1 where id = $2 returning *",
        )
        .bind(processed_at)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_currency(
        &self,
        ctx: &Context<'_>,
        currency: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set currency = $1 where id = $2 returning *",
        )
        .bind(currency)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_exchange_rate(
        &self,
        ctx: &Context<'_>,
        exchange_rate: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set exchange_rate = $1 where id = $2 returning *",
        )
        .bind(exchange_rate)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_fees(
        &self,
        ctx: &Context<'_>,
        fees: Option<f64>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set fees = $1 where id = $2 returning *",
        )
        .bind(fees)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_notes(
        &self,
        ctx: &Context<'_>,
        notes: Option<String>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set notes = $1 where id = $2 returning *",
        )
        .bind(notes)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn update_payment_processed_by_user_id(
        &self,
        ctx: &Context<'_>,
        processed_by_user_id: Option<Uuid>,
        id: Uuid,
    ) -> async_graphql::Result<payments::Model> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query_as::<_, payments::Model>(
            "update billing.payments set processed_by_user_id = $1 where id = $2 returning *",
        )
        .bind(processed_by_user_id)
        .bind(id)
        .fetch_one(db)
        .await?;

        Ok(result)
    }

    async fn remove_payment(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        let db = ctx.data::<sqlx::PgPool>()?;

        let result = sqlx::query("delete from billing.payments where id = $1")
            .bind(id)
            .execute(db)
            .await?;

        if result.rows_affected() != 1 {
            return Err(async_graphql::Error::new("Unable to remove payment"));
        }

        Ok("Payment removed successfully".into())
    }
}
